sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -2162,6 +2163,48 @@ object RemoveLiteralFromGroupExpressions extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Prunes unnecessary fields from a [[Generate]] if it is under a project which does not refer
+ * any generated attributes, .e.g., count-like aggregation on an exploded array.
+ */
+object GenerateOptimization extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformDownWithPruning(
+      _.containsAllPatterns(PROJECT, GENERATE), ruleId) {
+      case p @ Project(_, g: Generate) if p.references.isEmpty
+          && g.generator.isInstanceOf[ExplodeBase] =>
+        g.generator.children.head.dataType match {
+          case ArrayType(StructType(fields), _) =>
+            val atomicFields = fields.collect {
+              case f: StructField if f.dataType.isInstanceOf[AtomicType] => f
+            }
+            val extractor = if (atomicFields.size > 0) {
+              // Pick an arbitrary atomic field, if any
+              ExtractValue(g.generator.children.head,
```

Comment:
nit: I feel it's safer to create `GetStructField` instead of doing name lookup again. It's possible that some dataframe-generated query plan has name conflicts in the struct, and `GetStructField` allows us to put the ordinal directly to avoid a name lookup.

Reply:
okay, updated to use `GetArrayStructFields`.